hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/MembershipNamenodeResolver.java

```diff
@@ -478,4 +476,27 @@ private List<MembershipState> getRecentRegistrationForQuery(
   public void setRouterId(String router) {
     this.routerId = router;
   }
+
+  /**
+   * Shuffle cache, to ensure that the current nn will not be accessed first next time.
+   *
+   *
+   * @param nsId name service id
+   * @param namenode namenode contexts
+   */
+  @Override
+  public synchronized void shuffleCache(String nsId, FederationNamenodeContext namenode) {
+    cacheNS.compute(Pair.of(nsId, false), (ns, namenodeContexts) -> {
+      if (namenodeContexts != null
+              && namenodeContexts.size() > 0
```

Comment:
isEmpty() and probably we want to check for > 1 as there's nothing to rotate.
We should actually do that outside, if there's 1 or less, don't do anything.

Reply:
The reason for not judging outside here is that we should ensure that get cache and modify cache are atomic, considering the following situation:

1. After **Thread1** judges that the cache is not empty, it gets the ns no active in the cache
2. **Thread2** (loadCache thread  )clear the cache
3. **Thread3** processes the client request and finds that the cache is empty, and updates the cache. At this time, the ns - in the cache has active nn
4. **Thread1** rotates the previously acquired data of non-active nn and writes it into the cache, causing ns in the cache to have no active nn

Reply:
It is true that >1 should be judged here instead of >0

Reply:
Guaranteed atomicity can ensure that the above 2 and 3 are recognized before modifying the cache